design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,147 @@
+
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+We propose the following change to address the limitations.
+
+### Terms
+
+- Device Discovery. A Pod that is given a device or directory path filter and use the filter to discover devices on the host. The Pod updates the device orchestration Configmap with the device paths and mark the orchestration status as "DeviceDiscovered". The Pod is created by Rook Operator and runs in a Kubernetes deployment on the storage node. This process is storage backend agnostic. 
+
+- Device Provisioner. A Pod that is given device or directory paths upon start and make backend specific storage types. For instance, the provisioner prepares OSDs for Ceph backend. It is a Kubernetes batch job and exits after the devices are prepared.
+
+### Create new OSDs
+| Sequence |Rook Operator | Device Discovery  | Device Provisioner   | Ceph OSD Deployment | 
+|---|---|---|---|---|
+| 1  | Read devices from cluster CRD and create an OSD Discovery deployment on each storage node  |   |   |
+| 2  | Watch raw device Configmap  | Discover devices that fit those defined in device filters and update raw device Configmap with the device paths and filtering results| | |
+| 3  | Detect raw device Configmap change, parse Configmap, extract device paths with matching filtering results, and create an Device Provisioner deployment for each device  || | |
+| 4  | Watch device provisioning Configmap | | Prepare OSDs, Persist OSD ID, datapath, and node info in a per node device provisioning Configmap | |
+| 5  | Detect device provisioning Configmap change, parse Configmap, extract OSD info, construct OSD Pod command and arg | | | |
+| 6  | Create one deployment per OSD | | |
+| 7  | | | | Start `ceph-osd` Daemon one Pod per device |
+
+This change addresses the above limitations in the follow ways:
+- High reliability. Each `ceph-osd` daemon runs its own Pod, their restart and upgrade are by Kubernetes controllers. Upgrading Device Provisioner Pod no longer restarts `ceph-osd` daemons.
+- More efficient resource requests. Once Device Discovery detects all devices, Rook Operator is informed of the topology and assigns appropriate resources to each Ceph OSD deployment.
+- Reusable. Device discovery can be used for other storage backends.
+
+
+### Detailed Device Discovery Process
+
+Each `Device Discovery` Pod reads device filters from `Device Filter Configmap`, retrieves named device filters, walks through device trees and uses device filters to identify the devices, and stores the filtering results in `Raw Device Configmap`. 
+
+#### Device Filter Configmap
+
+This Configmap defines the node regex and device filters on each node. 
+
+A sample of filter Configmap is as the following:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: rook-system
+  name: device-filters
+data:
+  config: |
+    - nodes: "foo*" # node name regex
```

Comment:
this type of filter based on nodes and device names we already have in the cluster crd. Let's keep it in the crd instead of moving it to this separate filter configmap. But we do need to improve it to look at device size, etc.

Reply:
Agree CRD is the only user visible API for device filtering. The configmap is not exposed to end users. It is an internal object created by Operator based on cluster CRD definition. If the CRD is updated, Operator updates the Configmap and Discovery Pods will pick up the new filters on the next polling. 

This configmap is designed for programming purpose: Discovery Pods read from the configmap and get the information to probe devices. These Pods don't have to watch and handle cluster CRDs. 

The existing filter we have in 0.7 works at either on all nodes or individual node. Using node name regex hopefully can aggregate nodes and reduce the CRD size.

Reply:
I'm not sure we have a common understanding of the filter. What I'm saying is:
- The filter is global like the operator. The filter should discover all available devices on storage nodes
- Each device discovered can be consumed by exactly one rook cluster
- The filter for which devices are consumed by a cluster is specified in the cluster crd. 
So the filter defines what is consumed, rather than what is discovered.